crates/wasi-common/src/table.rs

```diff
@@ -31,14 +31,14 @@ impl Table {
 
     /// Insert a resource at the next available index.
     pub fn push(&mut self, a: Box<dyn Any>) -> Result<u32, Error> {
+        // NOTE: The performance of this new key calculation could be very bad once keys wrap
+        // around.
+        if self.map.len() == u32::MAX as usize {
```

Comment:
Are descriptors 0/1/2 counted here?

Reply:
As a side note, it may be useful to:

* Have a way to set an upper limit to the number of entries
* Have a way to get the current set of keys in the table, or at least the count. This is super useful in order to debug descriptor leaks (not just in hostcalls, also in applications).

Reply:
The ctx constructor asserts that descriptors 0/1/2 are populated, so I believe they are counted, though in a different spot.

I agree those features sound useful, but I think they're out of scope for landing the initial version of this.